### ISSUE TITLE
Show stars and planets in AR

### ISSUE THREAD

**Issue by GuiSousa135** [kylecorry31/Trail-Sense#1809]
Offline spherical planetarium function within the "astronomy" tab of the application to assist the user in the identification and observation of planets, stars, constellations, satelites and other celestial bodies. In addition, extra features such as autocalibrage, gyroscope positioning/location identification, free navigation mode. A simple and open source application that adds this functionality as exemple is Sky Map.

![Screenshot_20230609-151647_Sky Map](https://github.com/kylecorry31/Trail-Sense/assets/55188286/40e95833-3d41-4eda-abbf-b73862f6c314)

![Screenshot_20230612-130546_Sky Map](https://github.com/kylecorry31/Trail-Sense/assets/55188286/5874364e-55c1-4d34-9cb8-1efc548d38d6)

![Screenshot_20230612-130558_Sky Map](https://github.com/kylecorry31/Trail-Sense/assets/55188286/6f161068-317b-45e7-8946-869bdc1a012f)

**Comment by kylecorry31** [comment]
Hi, thank you for the suggestion. 

I may consider adding some of this once #1802 is complete. I have held off on adding star positions to Trail Sense because it's very hard to spot them using just a list representation of azimuth / altitude.

Criteria for adding a star / planet:
- The star is easily visible with the naked eye and can be used for navigation (ex. Polaris)
- The star is part of a major constellation the people might want to spot while camping or could use for celestial navigation
- The planet must be easily visible with the naked eye

But I likely will not add several planets or minor stars - see #449 and #319 where I previously attempted this but backed it off due to complexity and lack of practical use for the core use cases of Trail Sense.

Stars
- [x] Top 50 brightest stars
- [x] Any navigation star not included in the top 50
- [ ] Any constellation star not included in the top 50 (constellation(s) name shown under start name when a star is focused)

Constellations
- [ ] Ursa Major (Big Dipper)
Not finished

Planets
- [x] Venus
- [x] Mars
- [x] Jupiter
- [x] Saturn

Other
- [x] Sun
- [x] Moon
- [x] Meteor showers

Settings
- [ ] Show hard to see stars/planets (find a better name for this - shows low magnitude stars/planets)
- [ ] Show under horizon

**Comment by kylecorry31** [comment]
Another example app: https://github.com/tengel/AndroidPlanisphere

**Comment by GuiSousa135** [comment]
As the focus of the planetarium will be celestial navigation, the first objects to think about are the main constellations and their respective stars that are used to identify North and South, direction, time and other location parameters. We have three types of constellations: austral (southern hemisphere), boreal (northern hemisphere) and equatorial constellations, (located near the Celestial Equator) like the constellations of the zodiac. Then, the planets and visible stars of the earth: Mercury, Venus, Mars, Jupiter, Saturn, Uranus, Neptune, Sun and Moon. Finally, it would be important to see some of the imaginary lines of support and reference: the terrestrial equator line, the observer's horizon line (based on the location of the cell phone and the gyroscope), perhaps latitude and longitude angle markings. The Sky Map application (also made in Kotlin) demonstrates some of these objects. Soon I will be listing here specifically the name of the stars and constellations.

![Screenshot_20230711-083247_Bromite](https://github.com/kylecorry31/Trail-Sense/assets/55188286/68ed4427-a6f2-4bb8-9d54-d8580244681c)
![Screenshot_20230711-083214_Bromite](https://github.com/kylecorry31/Trail-Sense/assets/55188286/88a38b86-5d10-40e6-9378-a879d74708d8)

**Comment by kylecorry31** [comment]
@GuiSousa135 Thanks! As for the planets, I'll probably only add Mercury, Venus, Mars, Jupiter, and Saturn since the others aren't visible to the naked eye. I may start with just the stars, because I'm not too sure how important the planets are to celestial navigation (more research needed) - and they would mainly just be for people to stargaze while camping.

I'm going to start a list in my other comment, which I'll update as I get more information.

**Comment by GuiSousa135** [comment]
next milestone, please.

**Comment by kylecorry31** [comment]
> next milestone, please.

Sorry, this will not be in the next milestone - it is a low priority to be implemented in Trail Sense given other planned features, so not likely this year either.

If you are interested in implementing this, let me know and I can provide more technical details.

**Comment by maverick74** [comment]
If you allow an opinion, here's what i would like to see included:

- [ ] All night objects that are visible with the naked eye (as mentioned above, even the small ones)

     that includes 
    - stars
    - planets
- [ ] Constellation's identification (connections/draws/lines that connect them)
- [ ] Uranus
- [ ] Neptune

Note: i would like the inclusion of Uranus and Neptune because, while not visible, it would be great to be able to tell where they are in camp fires, for the kids... I mean... it's just two extra object and we're able to have the (for now) complete solar system. I believe it makes sense...

oh... and i almost forgot: please also include Planet Nine!!! I'm sure Batygin and Brown will appreciate the help in discovering it - ahahah - you'll get to name it and you'll gain some extra users in the process XP

**Comment by kylecorry31** [comment]
> If you allow an opinion, here's what i would like to see included:
> 
> * * [ ] 
>     [ ]  All night objects that are visible with the naked eye (as mentioned above, even the small ones)
>     that includes
>     
>     * stars* planets[ ] 
>   [ ]  Constellation's identification (connections/draws/lines that connect them)
>   [ ] 
>   [ ]  Uranus
>   [ ] 
>   [ ]  Neptune
> 
> Note: i would like the inclusion of Uranus and Neptune because, while not visible, it would be great to be able to tell where they are in camp fires, for the kids... I mean... it's just two extra object and we're able to have the (for now) complete solar system. I believe it makes sense...
> 
> oh... and i almost forgot: please also include Planet Nine!!! I'm sure Batygin and Brown will appreciate the help in discovering it - ahahah - you'll get to name it and you'll gain some extra users in the process XP

Thank you for the suggestions. An argument against adding the non-naked eye visible planets is that it may cause a little confusion if using them for celestial navigation. I'll need to consider the implications of that, but I don't imagine it would be too much of much of an issue.

**Comment by kylecorry31** [comment]
Another option is to split this. This could be for adding the celestial navigation objects, and I can have additional objects behind a setting - so, best of both worlds.

**Comment by maverick74** [comment]
oh, i can totally live with that!!!

It's a great idea :)

This way we could have 
- a simplified version of the sky where you could filter just a few objects without any criteria other that navigation. It would allow to hide almost any object you do not see fit.
- and a version that comprises all the other options mentioned above

letting the choice in the hands of the user (with the first as default)

i really like it! :)

**Comment by kylecorry31** [comment]
If anyone is interested in helping out with this issue, here's some places that I could use help:

- Add constellations to [Sol](https://github.com/kylecorry31/sol/blob/main/src/main/kotlin/com/kylecorry/sol/science/astronomy/stars/Constellation.kt). For now, just the [officially recognized constellations](https://starchild.gsfc.nasa.gov/docs/StarChild/questions/88constellations.html). But if there are any asterisms that are popular (like the big dipper), I'm open to having them included as well. To help me review, please list out the sources where you got the stars/constellations.
- Add stars to [Trail-Sense-Earth-Model](https://github.com/kylecorry31/Trail-Sense-Earth-Model/blob/main/generate-star-model.py) and [Sol](https://github.com/kylecorry31/sol/blob/main/src/main/kotlin/com/kylecorry/sol/science/astronomy/stars/Star.kt) (generated from the earth model) (optional, once in the earth model I can easily add to Sol). The star must be visible to the naked eye (I'll use a magnitude + constellation filter to filter out the minor objects in the default view). If the star isn't in simbad and you need to [add overrides](https://github.com/kylecorry31/Trail-Sense-Earth-Model/blob/main/scripts/simbad.py), please specify where you got the override from so I can verify.

Planet motion is more complicated and I have some books that document the algorithms for that, so I'm not looking to have the contributed at this moment (plus lower priority since they aren't ideal for celestial navigation).

**Comment by kylecorry31** [comment]
I felt like doing some astronomy math today, so in the next release the clearly visible planets will show up in AR at night (venus, mars, jupiter, saturn). Support for mercury, uranus, and neptune is in place but won't be available until I add a setting to show additional objects.

**Comment by GuiSousa135** [comment]
That's great! The Astronomy tool is one of my favorite functions, AR helps a lot and is very good see this moving forward.

**Comment by maverick74** [comment]
Hello!!! 

I'm not being able to see the origin location of meteor showers (such as the perseids)!!!

Is it supposed to be like this?

Also, I believe it would be nice - maybe under that "extra objects" switch - to be able to enable the view of objects under the horizon line. The point is: I maybe looking for a star that is not visible yet but could be good to to have an idea where it may be appearing...

**Comment by kylecorry31** [comment]
> Hello!!!
> 
> I'm not being able to see the origin location of meteor showers (such as the perseids)!!!
> 
> Is it supposed to be like this?
> 
> Also, I believe it would be nice - maybe under that "extra objects" switch - to be able to enable the view of objects under the horizon line. The point is: I maybe looking for a star that is not visible yet but could be good to to have an idea where it may be appearing...

Hi, yes the meteor shower origin has not been implemented yet - see https://github.com/kylecorry31/Trail-Sense/issues/1809#issuecomment-1587806249 for the current status of this issue in main.

I will consider adding a "show under horizon" switch as well, thanks!

**Comment by kylecorry31** [comment]
> > Hello!!!
> > I'm not being able to see the origin location of meteor showers (such as the perseids)!!!
> > Is it supposed to be like this?
> > Also, I believe it would be nice - maybe under that "extra objects" switch - to be able to enable the view of objects under the horizon line. The point is: I maybe looking for a star that is not visible yet but could be good to to have an idea where it may be appearing...
> 
> Hi, yes the meteor shower origin has not been implemented yet - see [#1809 (comment)](https://github.com/kylecorry31/Trail-Sense/issues/1809#issuecomment-1587806249) for the current status of this issue in main.
> 
> I will consider adding a "show under horizon" switch as well, thanks!

And now meteor shower radiants are implemented (will be in tonight's nightly build) ± 10 days of the peak date, so in the northern hemisphere you should be able to see the Perseids radiant in AR until around August 23.

**Comment by kylecorry31** [comment]
Since this issue is a little large and composed of tasks of various priorities, I'm closing it out in favor of the following:

- Constellations: https://github.com/kylecorry31/Trail-Sense/issues/3090
- Low brightness object toggle: https://github.com/kylecorry31/Trail-Sense/issues/3091
- Objects below the horizon: https://github.com/kylecorry31/Trail-Sense/issues/3092

I think that's everything from this issue that is remaining, if I missed something let me know.

**Comment by maverick74** [comment]
@kylecorry31 out of curiosity: what's missing in Ursa Major (Big Dipper), why do you consider it "Not finished"?

**Comment by kylecorry31** [comment]
> [@kylecorry31](https://github.com/kylecorry31) out of curiosity: what's missing in Ursa Major (Big Dipper), why do you consider it "Not finished"?

The list of constellations is not finished, not Ursa Major

**Comment by maverick74** [comment]
> The list of constellations is not finished, not Ursa Major

Oh... because Ursa Major was not checked i thought there was something wrong with it

<img width="265" height="100" alt="Image" src="https://github.com/user-attachments/assets/4245265a-d8d1-4ce6-b485-08a2a93916c4" />

Thanks for clarifying :)
